### PR TITLE
Add non-uniform segment sizing switch

### PR DESCRIPTION
**Pull request by DivineDominion**
This changes `PreferencesStyle.segmentedControl` to get an associated value:

```swift
public enum PreferencesStyle {
	public enum SegmentSize {
		case fit, uniform
	}
	
	case toolbarItems
	case segmentedControl(size: SegmentSize)
}
```

With that, you can opt in to not use the widest segmented control size for all segments. I still think the uniform sizing looks best for 2 or 3 segments. But when you have more than 4 panes, the names and large insets become a problem and require very wide windows.

| Uniform (old) | Fitting (new) |
| ------ | ------ |
| Screenshot 2020-02-20 at 18 59 28 | Screenshot 2020-02-20 at 18 59 58 |

### CODE REVIEW

**Comment by sindresorhus** [comment]
Good idea. Can you update the readme too? Maybe include the above comparison table and mention when you would use each style.

File: Sources/Preferences/PreferencesStyle.swift [hunk 1]
```diff
@@ -1,6 +1,11 @@
 import Cocoa
 
 public enum PreferencesStyle {
+	public enum SegmentSize {
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesStyle.swift]: Doc comment?

File: Sources/Preferences/PreferencesStyle.swift [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
+		case fit
+		case uniform
+	}
+
 	case toolbarItems
-	case segmentedControl
+	case segmentedControl(size: SegmentSize)
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesStyle.swift]: Isn't it technically only the width that changes? Maybe it should be:

```suggestion
	case segmentedControl(width: SegmentWidth)
```
?

File: Sources/Preferences/PreferencesStyle.swift [hunk 3]
```diff
@@ -1,6 +1,11 @@
 import Cocoa
 
 public enum PreferencesStyle {
+	public enum SegmentSize {
+		case fit
+		case uniform
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesStyle.swift]: Have you considered a default `auto` size? Which would switch to fit if `uniform` cannot fit.

I don't think users ever want this:

Screenshot 2020-02-22 at 14 11 03

So it kinda makes sense to do the right behavior by default.

Comment (sindresorhus) [reply Sources/Preferences/PreferencesStyle.swift]: In `fit` mode, I could fit 5 segmented controls.

Screenshot 2020-02-22 at 14 13 22

What if the user needs more? Have you considered a mode that shrinks the text?

Kinda like the system about dialog:

Screenshot 2020-02-22 at 14 13 48

File: Sources/Preferences/PreferencesStyle.swift (hunk 2 above)
Comment (sindresorhus) [reply Sources/Preferences/PreferencesStyle.swift]: Although, if we decide to go for shrinking the text too, as mentioned in https://github.com/sindresorhus/Preferences/pull/47#discussion_r382893228, `size` does make sense.

**Comment by sindresorhus** [comment]
@DivineDominion Friendly bump. In case you missed my comments :)

File: Sources/Preferences/PreferencesStyle.swift (hunk 3 above)
Comment (DivineDominion) [reply Sources/Preferences/PreferencesStyle.swift]: I like the `.fitSmall` idea!

Recently, I've been thinking that we maybe should drop the `.uniform` size completely. I don't know if anyone actually prefers uniform sizes. 🤔

But _if_ someone really cares about uniform sizes, I'd like to think that they won't like the segment to change style to `.fit` for some of the tabs, and `.uniform` for others. The "right thing" here, in my opinion, would be to not encourage sloppy preference windows where the segments jiggle around when you switch the active page.

@sindresorhus WDYT?

**Comment by DivineDominion** [comment]
I wanted to jump in and implement `.fitSmall`, but cannot help out with this repo for a while until I get a device with Catalina. That's unfortunate :)

image

**Comment by sindresorhus** [comment]
> but cannot help out with this repo for a while until I get a device with Catalina. That's unfortunate :)

I increased the required Xcode version as I thought that Xcode version was required to compile the latest Swift syntax. I've tried decreasing it here: https://github.com/sindresorhus/Preferences/pull/47/commits/c79a490cf4a1c992e15d75b7ab9bf70464a2ba41 Not sure you'll be able to compile though.

File: Sources/Preferences/PreferencesStyle.swift (hunk 3 above)
Comment (sindresorhus) [reply Sources/Preferences/PreferencesStyle.swift]: > Recently, I've been thinking that we maybe should drop the .uniform size completely. I don't know if anyone actually prefers uniform sizes. 🤔

I like the uniform style and I don't think we should get rid of it. It looks better than "fit" when there are just a few buttons. For example, in one of my apps, I use uniform style:

Screenshot 2020-06-05 at 23 03 26

> But if someone really cares about uniform sizes, I'd like to think that they won't like the segment to change style to .fit for some of the tabs, and .uniform for others. The "right thing" here, in my opinion, would be to not encourage sloppy preference windows where the segments jiggle around when you switch the active page.

Good point. Forgot about my idea about "auto". I didn't consider panes having different size.

**Comment by DivineDominion** [comment]
I won't have time to work on app components until much later this year. What's your plan with this going forward?

**Comment by sindresorhus** [comment]
I moved it to an issue in case anyone else wants to work on it sooner: https://github.com/sindresorhus/Preferences/issues/62

**Pull request closed without merging**